File: android/core/src/main/kotlin/com/drawlesschess/core/Rules.kt
package com.drawlesschess.core

enum class StalematePolicy {
    TRAPPED_PLAYER_LOSES,
    TRAPPED_PLAYER_WINS,
}

enum class DeadPositionPolicy {
    MATERIAL_VICTORY,
    FINAL_CAPTURE_VICTORY,
}

enum class BareKingPolicy {
    CONTINUE,
    BARE_KING_LOSES,
}

enum class FiftyMovePolicy {
    DISABLED,
    COMPLETING_PLAYER_LOSES,
    FORCED_MOVE_EXCEPTION,
    MATERIAL_VICTORY,
}

enum class EndReason {
    CHECKMATE,
    STALEMATE,
    REPETITION,
    DEAD_POSITION_MATERIAL,
    DEAD_POSITION_FINAL_CAPTURE,
    BARE_KING,
    FIFTY_MOVE_LIMIT,
    RESIGNATION,
    TIMEOUT,
}

data class MaterialValues(
    val pawn: Int = 1,
    val knight: Int = 3,
    val bishop: Int = 3,
    val rook: Int = 5,
    val queen: Int = 9,
) {
    init {
        require(pawn == 1 && knight == 3 && bishop == 3 && rook == 5 && queen == 9) {
            "Rules contract v1 requires standard material values"
        }
    }
}

data class RulesContractV1(
    val preset: Preset,
    val stalemate: StalematePolicy,
    val deadPosition: DeadPositionPolicy,
    val fiftyMove: FiftyMovePolicy,
    val repetitionThreshold: Int = 3,
    val completingPlayerLosesRepetition: Boolean = true,
    val forcedRepetitionException: Boolean = true,
    val materialValues: MaterialValues = MaterialValues(),
    val bareKing: BareKingPolicy = BareKingPolicy.BARE_KING_LOSES,
) {
    val schemaVersion: Int = 1

    init {
        require(repetitionThreshold == 3) { "Rules contract v1 fixes repetition at three occurrences" }
        require(completingPlayerLosesRepetition) { "Rules contract v1 requires completing-player loss" }
        require(forcedRepetitionException) { "Rules contract v1 requires the forced-move exception" }
        require(
            (preset == Preset.DRAWLESS && stalemate == StalematePolicy.TRAPPED_PLAYER_LOSES) ||
                (preset == Preset.ESCAPE && stalemate == StalematePolicy.TRAPPED_PLAYER_WINS),
        ) { "Preset and stalemate policy disagree" }
    }

    enum class Preset { DRAWLESS, ESCAPE }

    companion object {
        fun drawless(
            deadPosition: DeadPositionPolicy = DeadPositionPolicy.MATERIAL_VICTORY,
            fiftyMove: FiftyMovePolicy = FiftyMovePolicy.MATERIAL_VICTORY,
        ) = RulesContractV1(
            preset = Preset.DRAWLESS,
            stalemate = StalematePolicy.TRAPPED_PLAYER_LOSES,
            deadPosition = deadPosition,
            fiftyMove = fiftyMove,
        )

        fun escape(
            deadPosition: DeadPositionPolicy = DeadPositionPolicy.MATERIAL_VICTORY,
            fiftyMove: FiftyMovePolicy = FiftyMovePolicy.MATERIAL_VICTORY,
        ) = RulesContractV1(
            preset = Preset.ESCAPE,
            stalemate = StalematePolicy.TRAPPED_PLAYER_WINS,
            deadPosition = deadPosition,
            fiftyMove = fiftyMove,
        )
    }
}

data class PositionFacts(
    val mover: Side,
    val legalMovesAfter: Int,
    val sideToMoveInCheck: Boolean,
    val positionOccurrenceCount: Int,
    val repetitionAvoidingAlternativesBeforeMove: Int,
    val halfmoveClockAfter: Int,
    val fiftyMoveAvoidingAlternativesBeforeMove: Int,
    val deadPositionAfter: Boolean,
    val moveWasCapture: Boolean,
    val materialAfter: MaterialScore,
    val lastCaptureBy: Side?,
)

data class GameOutcome(
    val winner: Side,
    val loser: Side = winner.opposite(),
    val reason: EndReason,
) {
    init {
        require(loser == winner.opposite()) { "Outcome winner and loser must be opposite sides" }
    }
}

class DrawlessAdjudicator {
    fun adjudicate(rules: RulesContractV1, facts: PositionFacts): GameOutcome? {
        val sideToMove = facts.mover.opposite()

        if (facts.legalMovesAfter == 0) {
            if (facts.sideToMoveInCheck) {
                return GameOutcome(facts.mover, reason = EndReason.CHECKMATE)
            }
            val winner = if (rules.stalemate == StalematePolicy.TRAPPED_PLAYER_LOSES) {
                facts.mover
            } else {
                sideToMove
            }
            return GameOutcome(winner, reason = EndReason.STALEMATE)
        }

        if (facts.positionOccurrenceCount >= rules.repetitionThreshold) {
            val loser = if (facts.repetitionAvoidingAlternativesBeforeMove == 0) {
                sideToMove
            } else {
                facts.mover
            }
            return GameOutcome(loser.opposite(), reason = EndReason.REPETITION)
        }

        if (rules.bareKing == BareKingPolicy.BARE_KING_LOSES) {
            val winner = when {
                facts.materialAfter.white == 0 && facts.materialAfter.black > 0 -> Side.BLACK
                facts.materialAfter.black == 0 && facts.materialAfter.white > 0 -> Side.WHITE
                else -> null
            }
            if (winner != null) return GameOutcome(winner, reason = EndReason.BARE_KING)
        }

        if (facts.deadPositionAfter) {
            if (rules.deadPosition == DeadPositionPolicy.FINAL_CAPTURE_VICTORY) {
                check(facts.moveWasCapture) {
                    "Final-capture adjudication requires the transition move to be a capture"
                }
                return GameOutcome(facts.mover, reason = EndReason.DEAD_POSITION_FINAL_CAPTURE)
            }
            val winner = when {
                facts.materialAfter.white > facts.materialAfter.black -> Side.WHITE
                facts.materialAfter.black > facts.materialAfter.white -> Side.BLACK
                else -> facts.mover
            }
            return GameOutcome(winner, reason = EndReason.DEAD_POSITION_MATERIAL)
        }

        if (facts.halfmoveClockAfter >= 100) {
            val winner = when (rules.fiftyMove) {
                FiftyMovePolicy.DISABLED -> null
                FiftyMovePolicy.COMPLETING_PLAYER_LOSES -> sideToMove
                FiftyMovePolicy.FORCED_MOVE_EXCEPTION -> {
                    if (facts.fiftyMoveAvoidingAlternativesBeforeMove == 0) facts.mover
                    else sideToMove
                }
                FiftyMovePolicy.MATERIAL_VICTORY -> when {
                    facts.materialAfter.white > facts.materialAfter.black -> Side.WHITE
                    facts.materialAfter.black > facts.materialAfter.white -> Side.BLACK
                    facts.lastCaptureBy != null -> facts.lastCaptureBy
                    facts.fiftyMoveAvoidingAlternativesBeforeMove == 0 -> facts.mover
                    else -> sideToMove
                }
            }
            if (winner != null) {
                return GameOutcome(winner, reason = EndReason.FIFTY_MOVE_LIMIT)
            }
        }

        return null
    }
}

File: src/main/js/rules.js
export const Side = Object.freeze({ WHITE: "WHITE", BLACK: "BLACK" });
export const StalematePolicy = Object.freeze({
  TRAPPED_PLAYER_LOSES: "TRAPPED_PLAYER_LOSES",
  TRAPPED_PLAYER_WINS: "TRAPPED_PLAYER_WINS",
});
export const DeadPositionPolicy = Object.freeze({
  MATERIAL_VICTORY: "MATERIAL_VICTORY",
  FINAL_CAPTURE_VICTORY: "FINAL_CAPTURE_VICTORY",
});
export const FiftyMovePolicy = Object.freeze({
  DISABLED: "DISABLED",
  COMPLETING_PLAYER_LOSES: "COMPLETING_PLAYER_LOSES",
  FORCED_MOVE_EXCEPTION: "FORCED_MOVE_EXCEPTION",
});
export const EndReason = Object.freeze({
  NONE: "NONE",
  CHECKMATE: "CHECKMATE",
  STALEMATE: "STALEMATE",
  REPETITION: "REPETITION",
  DEAD_POSITION_MATERIAL: "DEAD_POSITION_MATERIAL",
  DEAD_POSITION_FINAL_CAPTURE: "DEAD_POSITION_FINAL_CAPTURE",
  FIFTY_MOVE_LIMIT: "FIFTY_MOVE_LIMIT",
});

export const opposite = (side) => side === Side.WHITE ? Side.BLACK : Side.WHITE;

export function ruleset({
  id = "drawless",
  stalemate = StalematePolicy.TRAPPED_PLAYER_LOSES,
  deadPosition = DeadPositionPolicy.MATERIAL_VICTORY,
  fiftyMove = FiftyMovePolicy.DISABLED,
} = {}) {
  return Object.freeze({ id, stalemate, deadPosition, fiftyMove });
}

export function positionAfterMove(overrides = {}) {
  const p = {
    mover: Side.WHITE,
    legalMoveCount: 1,
    sideToMoveInCheck: false,
    positionOccurrenceCount: 1,
    repetitionAvoidingAlternativesBeforeMove: 1,
    halfmoveClock: 0,
    fiftyMoveAvoidingAlternativesBeforeMove: 1,
    deadPosition: false,
    moveWasCapture: false,
    whiteMaterial: 0,
    blackMaterial: 0,
    ...overrides,
  };
  for (const key of ["legalMoveCount", "repetitionAvoidingAlternativesBeforeMove",
    "halfmoveClock", "fiftyMoveAvoidingAlternativesBeforeMove", "whiteMaterial", "blackMaterial"]) {
    if (!Number.isInteger(p[key]) || p[key] < 0) throw new RangeError(`${key} must be a non-negative integer`);
  }
  if (!Number.isInteger(p.positionOccurrenceCount) || p.positionOccurrenceCount < 1) {
    throw new RangeError("positionOccurrenceCount must be a positive integer");
  }
  if (![Side.WHITE, Side.BLACK].includes(p.mover)) throw new TypeError("mover must be WHITE or BLACK");
  return Object.freeze(p);
}

const ongoing = () => Object.freeze({ terminal: false, reason: EndReason.NONE });
const win = (reason, winner, explanation) => Object.freeze({
  terminal: true, reason, winner, loser: opposite(winner), explanation,
});

export function adjudicate(rules, p) {
  const sideToMove = opposite(p.mover);

  if (p.legalMoveCount === 0) {
    if (p.sideToMoveInCheck) {
      return win(EndReason.CHECKMATE, p.mover, `${p.mover} wins by checkmate`);
    }
    const winner = rules.stalemate === StalematePolicy.TRAPPED_PLAYER_LOSES
      ? p.mover : sideToMove;
    return win(EndReason.STALEMATE, winner, `${winner} wins under the ${rules.id} stalemate rule`);
  }

  if (p.positionOccurrenceCount >= 3) {
    const loser = p.repetitionAvoidingAlternativesBeforeMove === 0 ? sideToMove : p.mover;
    return win(EndReason.REPETITION, opposite(loser), `${loser} loses by causing a third repetition`);
  }

  if (p.deadPosition) {
    if (rules.deadPosition === DeadPositionPolicy.FINAL_CAPTURE_VICTORY) {
      if (!p.moveWasCapture) throw new Error("Final-capture adjudication requires a capture transition");
      return win(EndReason.DEAD_POSITION_FINAL_CAPTURE, p.mover,
        `${p.mover} wins by making the final meaningful capture`);
    }
    const winner = p.whiteMaterial > p.blackMaterial ? Side.WHITE
      : p.blackMaterial > p.whiteMaterial ? Side.BLACK : p.mover;
    return win(EndReason.DEAD_POSITION_MATERIAL, winner,
      `${winner} wins the dead position by material adjudication`);
  }

  if (p.halfmoveClock >= 100 && rules.fiftyMove !== FiftyMovePolicy.DISABLED) {
    const forced = rules.fiftyMove === FiftyMovePolicy.FORCED_MOVE_EXCEPTION
      && p.fiftyMoveAvoidingAlternativesBeforeMove === 0;
    const loser = forced ? sideToMove : p.mover;
    return win(EndReason.FIFTY_MOVE_LIMIT, opposite(loser),
      `${loser} loses by reaching the configured 50-move limit`);
  }

  return ongoing();
}
